src/main/scala/components/memory-ports.scala

```diff
@@ -0,0 +1,229 @@
+ *           (or things will possibly break) 
+ *   Output: request, a DecoupledIO that delivers a request from a memory port to memory. This is primarily
+ *           meant for connecting to an AsynMemIO's request input, and should not be connected to anything else
+ */
+class MemPortIO(val blockwidth: Int) extends Bundle {
+  // Pipeline <=> Port
+  val address  = Input(UInt(32.W))
+  val valid    = Input(Bool())
+  val ready    = Output(Bool())
+
+  // Port <=> Memory 
+  val response = Flipped(Valid(new Response(blockwidth)))
+  val request  = Decoupled(new Request(blockwidth))
+}
+
+/** 
+ * The *interface* of the IMemPort module.
+ *
+ * Pipeline <=> Port:
+ *   Input:  address, the address of an instruction in memory 
+ *   Input:  valid, true when the address specified is valid
+ *   Output: instruction, the requested instruction
+ *   Output: ready, true when memory is idling and ready for a request
+ */
+class IMemPortIO(override val blockwidth: Int) extends MemPortIO(blockwidth) {
+  val instruction = Output(UInt(blockwidth.W))
+}
+
+/**
+ * The *interface* of the DMemPort module.
+ *
+ * Pipeline <=> Port:
+ *   Input:  address, the address of a piece of data in memory. 
+ *   Input:  writedata, valid interface for the data to write to the address
+ *   Input:  valid, true when the address (and writedata during a write) specified is valid
+ *   Input:  memread, true if we are reading from memory
+ *   Input:  memwrite, true if we are writing to memory
+ *   Input:  maskmode, mode to mask the result. 0 means byte, 1 means halfword, 2 means word
+ *   Input:  sext, true if we should sign extend the result
+ *   Output: readdata, the data read and sign extended
+ *   Output: ready, true when memory is idling and ready for a request
+ */
+class DMemPortIO(override val blockwidth: Int) extends MemPortIO(blockwidth) {
+  val writedata = Input(UInt(blockwidth.W))
+  val memread   = Input(Bool())
+  val memwrite  = Input(Bool())
+  val maskmode  = Input(UInt(2.W))
+  val sext      = Input(Bool())
+
+  val readdata  = Output(UInt(blockwidth.W))
+}
+
+/**
+ * The instruction memory port.
+ *
+ * The I/O for this module is defined in [[IMemPortIO]].
+ */
+class IMemPort(val blockwidth: Int) extends Module {
+  val io = IO (new IMemPortIO(blockwidth))
+  io := DontCare
+  io.request.valid  := false.B
+  io.ready          := io.request.ready
+
+  // When the backing memory is ready and the pipeline is supplying a high valid signal
+  when (io.valid && io.request.ready) {
+    val request = Wire(new Request(blockwidth))
+    request := DontCare
+    request.address      := io.address
+    request.operation    := Read
+
+    io.request.bits  := request
+    io.request.valid := true.B
+  } .otherwise {
+    io.request.valid := false.B
+  }
+
+  // When the memory is outputting a valid instruction
+  when (io.response.valid) {
+    io.instruction := io.response.bits.data
+  }
```

Comment:
I don't believe we have a mechanism to determine if the outputted instruction or read data is "good". But in any case, wouldn't the hazard unit handle this, e.g. when an invalid instruction is fetched and so the CPU has to raise a fetch_access exception? 

The only checks to determine if an address points to a valid location in memory right now is if the address doesn't exceed the maximum possible addressable place in memory, but otherwise the instruction port sees the entire memory as valid instructions, and same with the data port seeing the entire memory as valid locations to read data from or write data to.

Reply:
That's not quite what I meant. 

When the address is sent to the `IMemPort`, the data on the `instruction` wire isn't valid yet since we haven't accessed memory. We need a way to know when that data becomes valid and the decode stage can proceed.

Reply:
I think we can just feed through the memory's `response.valid` wire directly into a `good` output in the `IMemPortIO`, and similarly the same can be done for the data port.